state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -194,37 +235,50 @@ impl From<DispatchError> for BatchExecutionError {
 	}
 }
 
-impl CcmSwapOutput {
-	/// Returns Some of tuple (principal, gas) after swap is completed.
-	/// else return None
-	pub fn completed_result(self) -> Option<(AssetAmount, AssetAmount)> {
-		if self.principal.is_some() && self.gas.is_some() {
-			Some((self.principal.unwrap(), self.gas.unwrap()))
-		} else {
-			None
-		}
-	}
-}
-
-// Cross chain message, including information at different stages.
-#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo)]
-pub(crate) struct CcmSwap {
-	source_asset: Asset,
-	deposit_amount: AssetAmount,
-	destination_asset: Asset,
-	destination_address: ForeignChainAddress,
-	deposit_metadata: CcmDepositMetadata,
-	principal_swap_id: Option<SwapId>,
-	gas_swap_id: Option<SwapId>,
-}
-
 pub struct CcmSwapAmounts {
 	pub principal_swap_amount: AssetAmount,
 	pub gas_budget: AssetAmount,
 	// if the gas asset is different to the input asset, it will require a swap
 	pub other_gas_asset: Option<Asset>,
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo)]
+enum CcmSwapState {
+	PrincipalSwapScheduled {
```

Comment:
I'm not sure I agree. When writing a state machine there will always be some place in code where externally the state transition happened and while we are updating the state variable, it is "out of date" for a few lines of code, and it is OK. If you think of `process_swap_outcomes` as the function that handles state transition (which it is), then the state we read is "previous" state, and what we write is the "new" state, and then it makes a lot of sense.

On the other hand, renaming to `PrincipalSwap` etc makes the code less clear IMO, as then it is less obvious that this is a step in the process, and starts to look more like some info about the principal swap (raising further questions like why things like `gas_budget` are stored on the principle swap).

I used `Regular` because there is only one state, so I don't think of it as a state machiine really, but more like the swaps's type, and in that sense it is more similar to (and consistent with) `NetworkFee` and `IngressEgressFee`.

Reply:
sure not a strong opinion